From 1c9eb4dfb9dfe477ce4f879949e5d5ba02ecb646 Mon Sep 17 00:00:00 2001
From: Jon Simons <jon@jonsimons.org>
Date: Sat, 9 Sep 2017 20:11:47 -0700
Subject: [PATCH] config: fix memory leak with repeated opcodes

Fix a memory leak in the path where parsing returns early due
to seeing a repeated opcode.  A testcase is added which
demonstrates the leak and fix with valgrind.

Resolves CID 1374267.

Signed-off-by: Jon Simons <jon@jonsimons.org>
Signed-off-by: Lars Wendler <polynomial-c@gentoo.org>
---
diff --git a/src/config.c b/src/config.c
index 42148df7..25d64998 100644
--- a/src/config.c
+++ b/src/config.c
@@ -251,6 +251,7 @@ static int ssh_config_parse_line(ssh_session session, const char *line,
   opcode = ssh_config_get_opcode(keyword);
   if (*parsing == 1 && opcode != SOC_HOST && opcode != SOC_UNSUPPORTED && opcode != SOC_INCLUDE) {
       if (seen[opcode] != 0) {
+          SAFE_FREE(x);
           return 0;
       }
       seen[opcode] = 1;
-- 
2.15.1

